cheroot/test/test_server.py

```diff
@@ -236,3 +237,54 @@ def test_peercreds_unix_sock_with_lookup(peercreds_enabled_server):
         peercreds_text_resp = requests.get(unix_base_uri + PEERCRED_TEXTS_URI)
         peercreds_text_resp.raise_for_status()
         assert peercreds_text_resp.text == expected_textcreds
+
+
+def test_high_number_of_file_descriptors():
+    """Test the server does not crash with a high file-descriptor value.
+
+    This test should cause a server crash, as the server will try to
+    select() a file-descriptor higher than 1024.
+    """
+    # Get current resource limits to restore them later
+    (soft_limit, hard_limit) = resource.getrlimit(resource.RLIMIT_NOFILE)
+    increased_nofile_limit = 2048
+
+    # Create our server
+    httpserver = HTTPServer(
+        bind_addr=(ANY_INTERFACE_IPV4, EPHEMERAL_PORT), gateway=Gateway,
+    )
+    # Hoard a lot of file descriptors by opening and storing a lot of sockets
+    test_sockets = []
+
+    try:
+        # We have to increase the nofile limit above 1024
+        # Otherwise we see a 'Too many files open' error, instead of
+        # an error due to the file descriptor number being too high
+        resource.setrlimit(
+            resource.RLIMIT_NOFILE,
+            (increased_nofile_limit, hard_limit),
+        )
+
+        # Open a lot of file descriptors, so the next ones the server
+        # opens are higher than 1024
+        for i in range(1024):
+            test_sockets.append(socket.socket())
+        assert test_sockets[-1].fileno() >= 1024
+
+        httpserver.prepare()
+        # This will trigger a crash if select() is used in the implementation
+        httpserver.tick()
+        with socket.socket() as sock:
+            assert sock.fileno() >= 1024
+
+    finally:
+        # Close our open resources
+        for test_socket in test_sockets:
+            test_socket.close()
```

Comment:
I'm not completely sure that it's needed here because opening sockets makes the kernel dedicate file descriptors with sequential numbers. So when you close them, it shouldn't reuse lower identifiers anyway.
How about closing them where you open these descriptors and not having to hold a variable with them?

Reply:
I tried this initially - it is required in order to increment the file descriptor number. I'm not sure what the internal logic of how/when descriptors are reallocated, but removing the array of open sockets triggers the assertion you suggested:
```python
            try:
>               assert sock.fileno() >= 1024
E               AssertionError: assert 12 >= 1024
```
I also tried just holding the most recently opened socket at any one time (open a socket, open a second socket, discard the first socket, etc.), but this also hit the assertion.

Reply:
Looks like I was wrong. The file descriptors are reused after close.

Reply:
I'm happy to try some more things if you think there's a problem with the implementation, but this reliably reproduces the bug in a minimal way, and my knowledge of lower-level constructs isn't massive - I'd rather focus on the solution.

A bit of StackOverflowing suggest that file descriptors aren't always allocated in this monotonically increasing way:
- https://stackoverflow.com/questions/9376189/are-file-descriptors-for-linux-sockets-always-in-increasing-order/9376493
- https://stackoverflow.com/questions/24557357/why-only-the-lowest-available-file-descriptor-is-allocated-in-unix

Reply:
Yeah, I've done some reading that implies that fds are reused after their deallocation. I guess it's fine to keep the current approach. I'll proceed with other aspects of the review now...